docs/Training-On-Your-Data.md

```diff
@@ -0,0 +1,354 @@
+
+### Offline Preprocessing
+
+If you want, you can also preprocess your data offline, and then upload the preprocessed data to cloud storage.
+
+Levanter has a script that basically runs the same online preprocessing code, but doesn't do any training.
+You can run it like this:
+
+```bash
+python -m levanter.main.cache_dataset --config_path my_config.yaml
+```
+
+You can actually connect this to a Ray cluster, and use the cluster to do the preprocessing. This lets you
+use any of Ray's autoscaling features to scale up the preprocessing job.
+
+To do so:
+
+```bash
+python -m levanter.main.cache_dataset \
+    --config_path my_config.yaml \
+    --address <coordinator-address> \
+    --start_workers false \
+    --auto_start_cluster false
+```
+
+## Configuration
+
+Levanter uses [Draccus](https://github.com/dlwh/draccus) to configure training runs. It's a YAML-to-dataclass
+library that also supports argument parsing via argparse. A detailed guide to configuring Levanter is available
+in the [Configuration Guide](./Configuration-Guide.md).
+
+This section will cover the basics of configuring a training run.
+
+### TL;DR
+
+Here's a configuration for a 1.4B parameter model with reasonable values for everything:
+
+```yaml
+data:
+    train_urls:
+      - "gs://path/to/train.{1..32}.jsonl.gz" # TODO
+    validation_urls:
+      - "gs://path/to/valid.{1..4}.jsonl.gz" # TODO
+    cache_dir: "gs://path/to/cache"  # TODO
+    tokenizer: "gpt2"  # any HF tokenizer path, or GCS path to an HF tokenizer
+model:
+  type: gpt2
+  hidden_dim: 1536
+  num_heads: 24
+  num_layers: 48
+  seq_len: 1024
+  gradient_checkpointing: true
+  scale_attn_by_inverse_layer_idx: true
+trainer:
+  wandb:
+    project: "levanter" # TODO
+    tags: [ "openwebtext", "gpt2"]
+
+  mp: p=f32,c=bfloat16
+  num_train_steps: 100000  # TODO
+  train_batch_size: 512  # you may need to tune this or per_device_parallelism
+  per_device_parallelism: -1
+  per_device_eval_parallelism: 8
+
+  max_eval_batches: null # set to a number to limit eval batches. useful if your eval set is enormous
+
+  checkpointer:
+    base_path: "gs://path/to/checkpoints"  # TODO
+    save_interval: 15m
+    keep:
+      - every: 10000
+optimizer:
+  learning_rate: 1E-4
+  weight_decay: 0.1
+  min_lr_ratio: 0.1
+# if you want:
+hf_save_steps: 10000
+hf_save_path: "gs://path/to/hf/checkpoints"  # TODO
+hf_upload: null # set to an hf repo if you want to upload automatically. You need to have logged in to hf-cli
+```
+
+If you want a different model size or architecture, you can look at the config files in
+[levanter/config](https://github.com/stanford-crfm/levanter/tree/main/config).
+
+### Continued Pretraining
+
+Levanter supports starting from an HF pretrained model. To do so, you should set your config like this:
+
+```yaml
+model:
+  type: mpt
+initialize_from_hf: "mosaicml/mpt-7b@68e1a8e0ebb9b30f3c45c1ef6195980f29063ae2"
+use_hf_model_config: true
+```
+
+TODO: add instructions for llama including auth.
```

Comment:
In my case, I just pass my token as env variable. Would it be possible to set in the yaml?

```
WANDB_API_KEY=$WANDB_API_KEY HUGGING_FACE_HUB_TOKEN=$HF_TOKEN  levanter/infra/run.sh python3 ...
```

Reply:
yep! I just figured this out myself!